pkg/ruler/ruler.go

```diff
@@ -217,6 +215,12 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.RingCheckPeriod = 5 * time.Second
 }
 
+func (cfg *Config) RulesBackupEnabled() bool {
+	// If the replication factor is greater the  1, only the first replica is responsible for evaluating the rule,
+	// the rest of the replica will store the rule groups as backup only for API HA.
+	return cfg.Ring.ReplicationFactor > 1
```

Comment:
Does rules replication work when there is only 1 zone? I am wondering if we need to also check number of zones here

Reply:
If zone awareness is disabled then replication works and API HA also works.

If zone awareness is enabled with only 1 zone the replication still works in the sense that rulers in the same zone will still backup the rule group, but the API will not be HA anymore because there is only one zone so the backup is useless.

We can check the number of zones here so we return false if there is only one zone and zone awareness is enabled, but currently there is no easy way to count the number of zones other than calling `GetAllInstanceDescs` and recording the unique set of zones from the instances. Can we add a new method in the `ReadRing` interface to return the zones?

Reply:
I don't understand why API HA doesn't work when there is only 1 zone. Then rulers in the same zone shouldn't backup rules. Or the backup rules need to be queried to achieve HA.
It should be an all or nothing thing.

Reply:
Its because the ring's `Get` method still returns x number of rulers even if there is only 1 AZ. But for the API uses the min between the number of zones and the replication when checking the replica which would be 1 in this case https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler_ring.go#L148 so it won't be able to tolerate any instance going down.

> It should be an all or nothing thing.
I agree. We can disable the backup if only 1 zone is available. Do you think we can add the method in the `ReadRing` to get the number of zones?

Reply:
Actually because of how ring's `Get` method work, in list rules we can consider that zone awareness is disabled with there is only 1 zone.

Reply:
Updated @yeya24 , ptal . Thank you